src/metabase/pulse/render/body.clj

```diff
@@ -397,8 +401,38 @@
                                    (tru "N/A")
                                    (format-percentage (/ value total)))]))}))
 
+(defn- donut-legend
+  [legend-entries]
+  (letfn [(table-fn [entries]
+            (into [:table {:style (style/style {:color       "#4C5773"
+                                                :font-family "Lato, sans-serif"
+                                                :font-size   "24px"
+                                                :font-weight "bold"
+                                                :box-sizing  "border-box"
+                                                :white-space "nowrap"})}]
+                  (for [{:keys [label percentage color]} entries]
+                    [:tr {:style (style/style {:margin-right "12px"})}
+                     [:td {:style (style/style {:color         color
+                                                :padding-right "7px"
+                                                :line-height   "0"})}
+                      [:span {:style (style/style {:font-size   "2.875rem"
+                                                   :line-height "0"
+                                                   :position    "relative"
+                                                   :top         "-4px"})} "•"]]
+                     [:td {:style (style/style {:padding-right "30px"})}
+                      label]
+                   [:td percentage]])))]
+    (if (< (count legend-entries) 8)
+      (table-fn legend-entries)
+      [:table (into [:tr]
+                    (map (fn [some-entries]
+                           [:td {:style (style/style {:padding-right  "20px"
+                                                      :vertical-align "top"})}
+                            (table-fn some-entries)])
+                         (split-at (/ (count legend-entries) 2) legend-entries)))])))
+
```

Comment:
The donut legend is given a list of entries which are expected to look like `{:label "some label, timestamp or any string" :percentage "42%" :color "#004412"}`. I pulled this out into its own function to keep the :categorical/donut method a bit easier to follow.

One thing of note in this function is an (arbitrary) choice to split the legend into 2 columns when the count is greater than 8 entries, where the list is just split in half. So, this means a legend of 8 entries will render in a single column, but a legend of 9 entries will render in 2 columns (the first has 5 entries).

Since that number is arbitrary, it can be changed if needed.

Additionally, I didn't put any logic into cases where 3 columns might make sense, its 1 or 2, and that's it. Not certain it's a perfect solution, but seems reasonable to me.

Reply:
I've also attempted to adjust the styling of the elements here (used a table to get better alignment when label names have variable width), but I'm not 100% happy with the 'magic number' nature of the pixel values here. Would be open to suggestions on cleanup here.

Reply:
I'm surprised to see nested tables as a solution. Can we not express this in a single table? `table-fn` returns a table and all of this is nested inside of :table > :tr > :td

Reply:
A nested table occurs when the legend is split into 2 'columns'. If there are 8 or fewer legend entries, it doesn't actually nest a table, just puts the table-fn table in directly.

Doing it this way let me adjust the padding between each column fairly easily. It was also the first solution I thought of, so if you think I could do it better, that's cool. I don't think this nesting is particularly bad or anything, though.